source/dhtslib/threadpool.d

```diff
@@ -0,0 +1,482 @@
+/**
+This module provides multithreading support through htslib. It allows for mapping delagates,
+and functions via a range interface with `parallelMap`. Crucially, the htslib task pool performs
+all work in order. All functions and interfaces below should as well.
+*/
+module dhtslib.threadpool;
+import dhtslib.memory;
+import htslib.hts;
+import htslib.thread_pool;
+import std.algorithm : map;
+import std.parallelism : totalCPUs;
+import std.traits;
+import std.range;
+import std.meta;
+import core.stdc.stdlib : calloc, free;
+import core.atomic : atomicOp;
+
+/// global thread pool
+__gshared ThreadPool globalPool;
+
+/// ensure global pool is initialized
+/// NOTE: only sets threads if the global pool is not already initialized
+void enforceGlobalThreadPool(int threads = -2) @trusted nothrow @nogc 
+{
+    if(globalPool.tpool && (globalPool.threads == threads || threads == -2)) {
+        return;
+    } else
+        globalPool = ThreadPool(threads, true);   
+}
+
+private template noUnsharedAliasing(T)
+{
+    enum bool noUnsharedAliasing = !hasUnsharedAliasing!T;
+}
+
+// This template tests whether a function may be executed in parallel from
+// @safe code via Task.executeInNewThread().  There is an additional
+// requirement for executing it via a TaskPool.  (See isSafeReturn).
+private template isSafeTask(F)
+{
+    enum bool isSafeTask =
```

Comment:
It's just another form of eponymous templates. It is used as compile-time logic like most of the functions in `std.traits`. So I could evaluate at compile-time the task to be executed is `@safe`. So this would mainly be used in like a `static if` statement.